target_chains/ton/contracts/contracts/Pyth.fc

```diff
@@ -254,8 +254,25 @@ int apply_decimal_expo(int value, int expo) {
     return result;
 }
 
-() execute_upgrade_contract(slice payload) impure {
-    ;; TODO: Implement
+() execute_upgrade_contract(cell new_code) impure {
+    load_data();
+    int hash_code = cell_hash(new_code);
+    throw_unless(ERROR_INVALID_CODE_HASH, upgrade_code_hash == hash_code);
+
+    ;; Set the new code
+    set_code(new_code);
+
+    ;; Set the code continuation to the new code
+    set_c3(new_code.begin_parse().bless());
+
+    ;; Throw an exception to end the current execution
+    ;; The contract will be restarted with the new code
+    throw(0);
```

Comment:
can you explain this a little bit? why not a normal return? so whenever there is a "throw" the state changes (like the code of a contract) really persist? in that case we should be very careful with our state changes.

Reply:
yup, in TON, when an exception is thrown, all changes made to the contract's persistent storage (c4) are preserved so as you can see in our other functions we only call `store_data` at the end of the functions to store to c4 after all the checks are done